lib/Doctrine/ORM/Query/Filter/DefaultFilterFactory.php

```diff
@@ -0,0 +1,55 @@
+<?php
+/*
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ * This software consists of voluntary contributions made by many individuals
+ * and is licensed under the MIT license. For more information, see
+ * <http://www.doctrine-project.org>.
+ */
+
+namespace Doctrine\ORM\Query\Filter;
+
+use Doctrine\ORM\Configuration;
+use Doctrine\ORM\EntityManagerInterface;
+
+class DefaultFilterFactory implements FilterFactoryInterface
+{
+    /**
+     * @var EntityManagerInterface
+     */
+    private $entityManager;
+
+    /**
+     * @var Configuration
+     */
+    private $ormConfiguration;
+
+    public function __construct(Configuration $ormConfiguration)
+    {
+        $this->ormConfiguration = $ormConfiguration;
+    }
+
+    public function setEntityManager(EntityManagerInterface $entityManager)
+    {
+        $this->entityManager = $entityManager;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function createFromName($name)
+    {
+        $filterClass = $this->ormConfiguration->getFilterClassName($name);
+        return new $filterClass($this->entityManager);
```

Comment:
That is what I meant yes. I would argue that there is a contract. Various parts of doctrine expect to be able to call `addFilterConstraint`, and this is only defined as an abstract method on `SQLFilter`. For example, `FilterCollection` has [return type annotations](https://github.com/doctrine/doctrine2/pull/1453/files#diff-ed7c8c396c80efdf4c0add7f3750a213L88) which specify that `SQLFilter` instances are returned.

If it's the case that the EM is not required for some filters, it would be cleaner imo to create a `FilterInterface` which just specifies `addFilterConstraint` - however any time you want to use parameters you will need the EM in order to quote them so this still doesn't solve the core problem that an EM is often required.